modalc-lib/modal-vector.rkt

```diff
@@ -0,0 +1,118 @@
+#lang racket
+
+(require modalc
+         syntax/parse/define)
+
+(define (modal-vector/c should-apply-ctc? . inner-ctc)
+  (define inner-ctc-proj
+    (map contract-late-neg-projection inner-ctc))
+  (make-contract
+   #:name `(modal-vector/c ,(map contract-name inner-ctc-proj))
+   #:late-neg-projection
+   (λ (blame)
+     (define inner-ctc-proj/blame (map (λ (p) (p (blame-swap blame))) inner-ctc-proj))
+     (λ (val neg-party)
+       (impersonate-vector val
+                           (λ (vec index value)
+                             (cond [(should-apply-ctc? (list index value))
+                                    (printf "ref call: ~s\n" (list index value))
+                                    (for/list ([arg (in-vector vec)]
+                                               [proj (in-list inner-ctc-proj/blame)])
+                                      (proj arg neg-party))
+                                    value]
+                                   [else (printf "no ref ~s\n" (list index value))
+                                         value]))
+                           (λ (vec index value)
+                             (cond [(should-apply-ctc? (list index value))
+                                    (printf "set call: ~s\n" (list index value))
+                                    (for/list ([arg (in-vector vec)]
+                                               [proj (in-list inner-ctc-proj/blame)])
+                                      (proj arg neg-party))
+                                    (vector-set! vec index value)
+                                    value]
+                                   [else (printf "no set ~s\n" (list index value))
+                                         value])))))))
```

Comment:
Since we're going down the `modal-vector/c*` route, this one really should just use it instead of duplicating most of the code.

Reply:
Sorry, I just made a copy to make edits with and I just left it there. This wasn't meant to be a double-implementation.